### DISCUSSION TITLE
What would be the correct way to work with array methods in the RunJavaScript activity?

### DISCUSSION THREAD

**Discussion by RainerAtSpirit** [elsa-workflows/elsa-core#2446] [Q&A] [answered]
Here's a simple console use case:

```javascript
let TestArray = [{test: 123}]
TestArray.push({"test": 123})
TestArray = [...TestArray, {"test2": 234}]
TestArray = TestArray.concat([{"test3": 456}])
JSON.stringify(TestArray) // '[{"test":123},{"test":123},{"test2":234},{"test3":456}]'
```

When trying to do the same in a workflow, the `TestArray` variable won't get updated. 

![image](https://user-images.githubusercontent.com/581644/140048698-4aa2fc4c-f3f1-42b1-b537-d659aeadbf74.png)

**Definition:**

![image](https://user-images.githubusercontent.com/581644/140048780-4890d100-29e1-49cc-b96e-50efc910e8ff.png)

![image](https://user-images.githubusercontent.com/581644/140048808-7e223a1a-1a6e-4d88-98f3-61532e5fd1e0.png)

**DefinitionExport:**

[array-push-issue.zip](https://github.com/elsa-workflows/elsa-core/files/7467068/array-push-issue.zip)

**Comment by RainerAtSpirit** [comment, marked as answer]
Looks like mutating the global `TestArray` is not supported. However, creating a locally scoped variable works.

```javascript
let myArray = getVariable('TestArray')
myArray.push({"test": 123})
myArray = [...myArray, {"test2": 234}]
myArray = myArray.concat([{"test3": 456}])
setVariable('TestArray', myArray)
```

![image](https://user-images.githubusercontent.com/581644/140293473-3cd392d0-61f8-48f6-807c-f0917da5c036.png)

**Reply by sfmskywalker** [reply]
I didn't know that we can't update variables, but it probably makes sense to prevent scripts from changing values stored in the engine.
Good to see you found a way around it!

/cc @sebastienros @lahma

**Reply by lahma** [reply]
I'm afraid I don't understand the context well enough here. When running the example code in Jint REPL I get the expected output (like in NodeJS), so Jint works as expected *when doing raw JS*. There might be some interop or wrapping in play here that might change the behavior.

**Reply by sfmskywalker** [reply]
Hi @lahma , thanks for popping up!

The question I wanted to ask was basically like this:
Given the following code:

```csharp
var engine = new Engine();
engine.SetValue("Foo", "Bar);
var result = engine.Evaluate("Foo = 'Baz'").ToString();

// Will result be "Baz"?
```

I just tried this myself and the answer is in the affirmative: it just works.
For a second I thought maybe this would not be allowed, which would have explained why the following script would not have the desired effect (of actually updating the `TestArray` variable:

```javascript
TestArray.push({"test": 123})
TestArray = [...TestArray, {"test2": 234}]
TestArray = TestArray.concat([{"test3": 456}])
```

But even that works fine when I try it in isolation, so this is clearly not an issue with Jint 😃

**Reply by sfmskywalker** [reply]
I understand now. Although scripts can update variables, it is up to the hosting application to then read these values back (if they want to). In the case of Elsa, it is simply setting values before a script is executed, but it does not then read these values back.

**Reply by lahma** [reply]
One thing to experiment might be setting engine to always be in strict mode via `new Engine(options => options.Strict());` this is the same thing as having all scripts start with `"use strict";`. This will both improve performance and effectively disables JS's [sloppy mode](https://developer.mozilla.org/en-US/docs/Glossary/Sloppy_mode) where you can (among other bad things) assign to undeclared variables.

Of course this might surprise users if they are used to doing "bad JS".
